src/bayesdb.py

```diff
@@ -128,6 +142,49 @@ def sql_untrace(self, tracer):
         assert self.sql_tracer == tracer
         self.sql_tracer = None
 
+    def check_uncompleted_session_entries(self):
+        '''Check if the previous session ended with a failed command and
+        suggest sending the session'''
+        cursor = self.sqlite3.execute('''SELECT COUNT(*) FROM
+        bayesdb_session_entries WHERE completed=0 AND session_id=?''', (self.session_id-1,))
+        uncompleted_entries = int([row[0] for row in cursor][0])
+        if uncompleted_entries > 0:
+            print '''WARNING: Previous session contains uncompleted entries. This may be due to a bad termination or crash of the previous session. Consider uploading the session.'''
```

Comment:
Date: Wed, 30 Sep 2015 15:17:40 -0700
   From: gregory-marton <notifications@github.com>

   > @@ -128,6 +142,49 @@ def sql_untrace(self, tracer):
   >          assert self.sql_tracer == tracer
   >          self.sql_tracer = None
   >  
   > +    def check_uncompleted_session_entries(self):

Please don't add methods to the BayesDB class unless they're really
intended to be part of the public API for database handles.

   > +        '''Check if the previous session ended with a failed command and
   > +        suggest sending the session'''

First sentence of docstring must fit on one line.

   > +        cursor = self.sqlite3.execute('''SELECT COUNT(*) FROM
   > +        bayesdb_session_entries WHERE completed=0 AND session_id=?''', (self.session_id-1,))

Please follow bayeslite's convention for embedding and indenting SQL
queries.

cursor = self.sqlite3.execute('''
    SELECT COUNT(*) FROM bayesdb_session_entries
        WHERE completed = 0 AND session_id = ?
''', (self.session_id - 1,))

   > +        uncompleted_entries = int([row[0] for row in cursor][0])

Just say cursor.next()[0].  No need for int conversion -- if you don't
get an int that's a bug and should not be quietly suppressed.

   > +        if uncompleted_entries > 0:
   > +            print '''WARNING: Previous session contains uncompleted entries. This may be due to a bad termination or crash of the previous session. Consider uploading the session.'''

   Not sure what the right warning mechanism is, but printing to stdout is unlikely to be it. Let's ask @riastradh-probcomp ?

Printing to stdout is wrong.  There is currently no mechanism for
logging error/warning messages to a database handle, but there is an
obvious design for one based on sqlite3's error log -- user calls
bdb.set_error_hook(hook) or something and then messages will be passed
to the given hook function.

(Eventually this should hook into sqlite3's error log too, but that's
not expedient with Python's broken sqlite3 module.  Now that nobody
seems to care about the pain of bringing in external dependencies, we
should just switch to using Python apsw.  I have a branch from
February to do that.  May be worth reviving, may be better to start
from scratch.)